packages/flutter/lib/src/widgets/navigator.dart

```diff
@@ -2628,10 +2675,33 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
     }
     assert(!_debugLocked);
     assert(() { _debugLocked = true; return true; }());
+    // We have to manually extract the inherited widget in initState because
+    // the current context is not fully initialized.
+    final InheritedNavigatorObserver inheritedNavigatorObserver = context
+      .getElementForInheritedWidgetOfExactType<InheritedNavigatorObserver>()
+      ?.widget as InheritedNavigatorObserver;
+    _updateInheritedObserver(inheritedNavigatorObserver?.observer);
     _flushHistoryUpdates();
+    // Reset the reference so that it can be initialized in the
+    // didChangeDependencies.
+    _updateInheritedObserver(null);
     assert(() { _debugLocked = false; return true; }());
   }
 
+  @override
+  void didChangeDependencies() {
+    super.didChangeDependencies();
+    _updateInheritedObserver(InheritedNavigatorObserver.of(context));
+  }
+
+  void _updateInheritedObserver(NavigatorObserver inheritedObserver) {
```

Comment:
nit: rename the argument to `newObserver` to avoid the confusion between `inheritedObserver` and `_ inheritedObserver`?